### PR TITLE
[x/gamm][stableswap]: Create scaling/descaling functions for token inputs/outputs

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #2902

## What is the purpose of the change

This PR adds a scaling function for token inputs. Depending on what level we want to handle rounding at (at the scaling function level vs. at the `SwapExactAmountIn` level), we can either use this function for outputs as well or create a second one that handles outputs with different rounding.

I personally think having one function with rounding handled at a higher level of abstraction is cleaner because it has less repeat code, but I can see the security argument for having multiple scaling functions that round differently for each specific use (cc: @ValarDragon)

## Brief Changelog

- Add `scaledInput` function and tests

## Testing and Verifying

- The function added in this PR is tested in `pool_test.go` under `TestScaledInput`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

**Comment by AlpinYukseloglu** [comment]
Switched to draft to add descaling as well

File: x/gamm/pool-models/stableswap/pool.go [hunk 1]
```diff
@@ -132,6 +133,23 @@ func (p Pool) scaledPoolReserves() ([]sdk.DecCoin, error) {
 	return scaledReserves, nil
 }
 
+// scaledInput returns scaled input tokens for usage in AMM equations
+func (p Pool) scaledInput(input sdk.Coins) ([]sdk.DecCoin, error) {
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:137]:
```suggestion
func (p Pool) scaledInput(input sdk.Coin) (sdk.DecCoin, error) {
```

**Review by ValarDragon — APPROVED** [review]

**Comment by ValarDragon** [comment]
Should change division to method in #2908 once thats merged

File: x/gamm/pool-models/stableswap/pool.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/pool.go]: I originally had this as `sdk.Coins` to be able to accommodate inputs for all-asset joins as well – will leave it as `sdk.Coin` for now as it's simpler and aligned with the descaling function (which deals with 1 asset) and handle all asset joins at a higher level of abstraction

File: x/gamm/pool-models/stableswap/pool.go [hunk 2]
```diff
@@ -104,6 +105,22 @@ func (p Pool) NumAssets() int {
 	return len(p.PoolLiquidity)
 }
 
+// scaledInput returns scaled input tokens for usage in AMM equations
+func (p Pool) scaledInput(input sdk.Coin, roundingDirection osmomath.RoundingDirection) (sdk.DecCoin, error) {
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:109]:
```suggestion
func (p Pool) scaleInputAmount(input sdk.Coin, roundingDirection osmomath.RoundingDirection) (osmomath.BigDec, error) {
```

File: x/gamm/pool-models/stableswap/pool.go [hunk 3]
```diff
@@ -104,6 +105,22 @@ func (p Pool) NumAssets() int {
...
+	scaledAmount, err := osmomath.DivIntByU64ToBigDec(input.Amount, scalingFactor, roundingDirection)
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+	scaledInput := sdk.NewDecCoinFromDec(
+		input.Denom,
+		scaledAmount.SDKDec())
+
+	// TODO: correctly handle rounding
+	return scaledInput, nil
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:121]:
```suggestion
	if err != nil {
		return osmomath.BigDec{}, err
	}
	return scaledAmount, nil
```

**Pull request merged**